File: skills/runbook/tests/test_move_remove_step.py
from __future__ import annotations

import subprocess
import sys
import tempfile
import unittest
from pathlib import Path

from test_helpers import REFERENCE_TEMPLATE, RUNCTL, load_text

sys.path.insert(0, str(RUNCTL.parent))

import commands.validate as validate_cmd


class MoveRemoveStepTests(unittest.TestCase):
    maxDiff = None

    def setUp(self) -> None:
        self.template_text = load_text(REFERENCE_TEMPLATE)

    def test_runctl_remove_step_removes_plan_and_record_item(self) -> None:
        with tempfile.TemporaryDirectory() as tmpdir:
            runbook_path = Path(tmpdir) / "authority.md"
            runbook_path.write_text(self.template_text, encoding="utf-8")
            result = subprocess.run(
                [sys.executable, str(RUNCTL), "remove-step", str(runbook_path), "--item", "2"],
                capture_output=True,
                text=True,
                check=False,
            )

            self.assertEqual(0, result.returncode)
            content = runbook_path.read_text(encoding="utf-8")

        self.assertIn("[runbook-remove-step] removed item 2", result.stdout)
        self.assertNotIn("### 2. <编号项标题>", content)
        self.assertEqual([], validate_cmd.collect_errors(content))

    def test_runctl_move_step_reorders_existing_item(self) -> None:
        with tempfile.TemporaryDirectory() as tmpdir:
            runbook_path = Path(tmpdir) / "authority.md"
            runbook_path.write_text(self.template_text, encoding="utf-8")

            add_result = subprocess.run(
                [
                    sys.executable,
                    str(RUNCTL),
                    "add-step",
                    str(runbook_path),
                    "--title",
                    "收尾检查",
                ],
                capture_output=True,
                text=True,
                check=False,
            )
            self.assertEqual(0, add_result.returncode)

            move_result = subprocess.run(
                [
                    sys.executable,
                    str(RUNCTL),
                    "move-step",
                    str(runbook_path),
                    "--item",
                    "3",
                    "--after",
                    "1",
                ],
                capture_output=True,
                text=True,
                check=False,
            )

            self.assertEqual(0, move_result.returncode)
            content = runbook_path.read_text(encoding="utf-8")

        self.assertIn("[runbook-move-step] moved item 3 after 1", move_result.stdout)
        self.assertEqual(2, content.count("### 2. 收尾检查"))
        self.assertEqual(2, content.count("### 3. <编号项标题>"))
        self.assertEqual([], validate_cmd.collect_errors(content))


if __name__ == "__main__":
    unittest.main()

File: skills/runbook/tests/test_move_remove_step.mjs
import test from "node:test";
import assert from "node:assert/strict";
import os from "node:os";
import path from "node:path";
import { mkdtempSync, readFileSync, writeFileSync } from "node:fs";
import { rmSync } from "node:fs";
import { collectErrors } from "../scripts/commands/validate.mjs";
import { REFERENCE_TEMPLATE, loadText, runRunctl } from "./helpers.mjs";

test("runctl remove-step removes plan and record item", async () => {
  const dir = mkdtempSync(path.join(os.tmpdir(), "runbook-move-remove-"));
  try {
    const runbookPath = path.join(dir, "authority.md");
    writeFileSync(runbookPath, loadText(REFERENCE_TEMPLATE), "utf8");
    const result = await runRunctl(["remove-step", runbookPath, "--item", "2"]);
    assert.equal(result.status, 0);
    const content = readFileSync(runbookPath, "utf8");
    assert.match(result.stdout, /\[runbook-remove-step] removed item 2/);
    assert.ok(!content.includes("### 🔴 2. <编号项标题>"));
    assert.deepEqual(await collectErrors(content), []);
  } finally {
    rmSync(dir, { recursive: true, force: true });
  }
});

test("runctl move-step reorders existing item", async () => {
  const dir = mkdtempSync(path.join(os.tmpdir(), "runbook-move-remove-"));
  try {
    const runbookPath = path.join(dir, "authority.md");
    writeFileSync(runbookPath, loadText(REFERENCE_TEMPLATE), "utf8");
    const addResult = await runRunctl(["add-step", runbookPath, "--title", "收尾检查"]);
    assert.equal(addResult.status, 0);
    const moveResult = await runRunctl(["move-step", runbookPath, "--item", "3", "--after", "1"]);
    assert.equal(moveResult.status, 0);
    const content = readFileSync(runbookPath, "utf8");
    assert.match(moveResult.stdout, /\[runbook-move-step] moved item 3 after 1/);
    assert.equal((content.match(/### 🟡 2\. 收尾检查/g) ?? []).length, 2);
    assert.equal((content.match(/### 🔴 3\. <编号项标题>/g) ?? []).length, 2);
    assert.deepEqual(await collectErrors(content), []);
  } finally {
    rmSync(dir, { recursive: true, force: true });
  }
});
